docs/okeps/okep-5377-extend-udn-to-support-multiple-cluster-subnets-in-layer3-topology.md

```diff
@@ -0,0 +1,273 @@
+Changes are required in the `SecondaryNetworkClusterManager` to ensure stable node subnet assignments
+when new subnets are dynamically added.
+
+### OVN Network Topology
+
+Adding new subnets to a Layer3 network simply extends its cluster subnet allocation pool. Each node
+continues to allocate one subnet from the available pool. The overall OVN network topology remains
+largely unchanged. However, one additional change is required in the routing policies of the OVN
+Cluster Router: the `DefaultNoReroutePriority` (priority 102) policies.
+
+Currently it only allows east-west traffic within the individual cluster subnet.
+
+For example, for a network with cluster subnets `10.1.0.0/16` and `10.2.0.0/16`, the OVN cluster
+router policy list includes:
+```
+$ ovn-nbctl lr-policy-list  udn_udn.primary.layer3_ovn_cluster_router
+Routing Policies
+      ...
+      102   ip4.src == 10.1.0.0/16 && ip4.dst == 10.1.0.0/16    allow
+      102   ip4.src == 10.2.0.0/16 && ip4.dst == 10.2.0.0/16    allow
+```
+
+To enable east-west traffic between the subnets, the following additional policies must be added:
+```
+$ ovn-nbctl lr-policy-list  udn_udn.primary.layer3_ovn_cluster_router
+Routing Policies
+      ...
+      102   ip4.src == 10.1.0.0/16 && ip4.dst == 10.1.0.0/16    allow
+      102   ip4.src == 10.1.0.0/16 && ip4.dst == 10.2.0.0/16    allow
+      102   ip4.src == 10.2.0.0/16 && ip4.dst == 10.2.0.0/16    allow
+      102   ip4.src == 10.2.0.0/16 && ip4.dst == 10.1.0.0/16    allow
+```
+
+
+### API Validations
+
+The `subnets` field in the current CRD has the following validation rules:
+```
+	// +kubebuilder:validation:MaxItems=2
+	// +kubebuilder:validation:XValidation:rule="size(self) != 2 || !isCIDR(self[0].cidr) || !isCIDR(self[1].cidr) || cidr(self[0].cidr).ip().family() != cidr(self[1].cidr).ip().family()", message="When 2 CIDRs are set, they must be from different IP families"
+	Subnets []Layer3Subnet `json:"subnets,omitempty"`
+```
+
+* `MaxItems` is required to keep CEL evaluation costs below threshold. It has to be
+  retained but adjusted to a reasonable value (the max value is 400 for the new rules defined below).
+* The `XValidation:rule` must be removed to allow more than one subnet per IP family.
+
+New validation logic should be added to enforce the following constraints on `layer3.subnets`:
+* Prevent removal or modification of existing subnets:
+  ```yaml
+    - message: Removing existing subnets is not allowed
+      rule: '!has(oldSelf.subnets) || oldSelf.subnets.all(old, self.subnets.exists(new,
+        new.cidr == old.cidr))'
+  ```
+* Prevent overlapping or nested subnets:
+  ```yaml
+    - message: Subnets must not overlap or contain each other
+      rule: '!has(self.subnets) || self.subnets.size() == 1 || !self.subnets.exists(i,
+        self.subnets.exists(j, i != j && cidr(i.cidr).containsCIDR(j.cidr)))'
+  ```
+* Ensure all subnets use the same hostSubnet size:
```

Comment:
What is the maximum number of subnets that we should support per UDN?

Reply:
This hasn’t been decided yet. In practice, most users will use only one subnet, so it’s unlikely to be a concern.

Reply:
Yes, but since the UDN CRD is opinionated, we should provide a limit. I would recommend capping this at 16 to start with.

Reply:
IMO, there’s no need to decide on this number at this stage. Users who choose to use multiple subnets do so at their own risk.